src/client/dfuse/dfuse_cont.c

```diff
@@ -27,123 +28,79 @@ dfuse_cont_open(fuse_req_t req, struct dfuse_inode_entry *parent,
-			{
-				struct fuse_entry_param	entry = {0};
-
-				DFUSE_TRA_DEBUG(parent, "Checking %p", dfsi);
-
-				if (uuid_compare(dfsi->dfs_cont,
-						 dfs->dfs_cont) != 0)
-					continue;
-
-				DFUSE_TRA_INFO(parent,
-					       "Found existing container dfs %p",
-					       dfsi);
-
-				rc = dfuse_check_for_inode(fs_handle, dfsi,
-							   &ie);
-				if (rc == -DER_NONEXIST) {
-					D_FREE(dfs);
-					dfs = dfsi;
-					D_GOTO(alloc_ie, 0);
-				} else {
-					D_ASSERT(rc == -DER_SUCCESS);
-				}
-
-				DFUSE_TRA_INFO(ie,
-					       "Reusing existing container entry without reconnect");
-
-				/* Update the stat information, but copy in the
-				 * inode value afterwards.
-				 */
-				rc = dfs_ostat(ie->ie_dfs->dfs_ns, ie->ie_obj,
-					       &entry.attr);
-				if (rc) {
-					DFUSE_TRA_ERROR(ie,
-							"dfs_ostat() failed: (%s)",
-							strerror(rc));
-					D_GOTO(err_unlock, rc);
-				}
-
-				entry.attr.st_ino = ie->ie_stat.st_ino;
-				entry.generation = 1;
-				entry.ino = entry.attr.st_ino;
-				DFUSE_REPLY_ENTRY(ie, req, entry);
-				D_MUTEX_UNLOCK(&fs_handle->dpi_info->di_lock);
-				D_FREE(dfs);
-				return;
-			}
+			D_GOTO(err, rc);
 		}
 	}
 
-	rc = daos_cont_open(dfp->dfp_poh, dfs->dfs_cont,
-			    DAOS_COO_RW, &dfs->dfs_coh, NULL, NULL);
-	if (rc == -DER_NONEXIST) {
-		DFUSE_TRA_INFO(dfs, "daos_cont_open() failed: (%d)", rc);
-		D_GOTO(err_unlock, rc = daos_der2errno(rc));
-	} else if (rc != -DER_SUCCESS) {
-		DFUSE_TRA_ERROR(dfs, "daos_cont_open() failed: (%d)", rc);
-		D_GOTO(err_unlock, rc = daos_der2errno(rc));
-	}
+	rc = dfuse_cont_open(fs_handle, dfp, &cont, &dfc);
+	if (rc)
+		D_GOTO(err, rc);
 
-	rc = dfs_mount(dfp->dfp_poh, dfs->dfs_coh, O_RDWR, &dfs->dfs_ns);
-	if (rc) {
-		DFUSE_TRA_ERROR(ie, "dfs_mount() failed: (%s)", strerror(rc));
-		D_GOTO(close, rc);
-	}
+	rlink = d_hash_rec_find(&fs_handle->dpi_iet,
+				&dfc->dfs_ino,
+				sizeof(dfc->dfs_ino));
+	if (rlink) {
+		struct fuse_entry_param entry = {0};
 
-	d_list_add(&dfs->dfs_list, &dfp->dfp_dfs_list);
+		ie = container_of(rlink, struct dfuse_inode_entry, ie_htl);
+
+		DFUSE_TRA_INFO(ie,
+			       "Reusing existing container entry without reconnect");
+
+		/* Update the stat information, but copy in the
+		 * inode value afterwards.
+		 */
+		rc = dfs_ostat(ie->ie_dfs->dfs_ns, ie->ie_obj,
+			       &entry.attr);
+		if (rc) {
+			DFUSE_TRA_ERROR(ie,
+					"dfs_ostat() failed: (%s)",
+					strerror(rc));
+			D_GOTO(decref, rc);
+		}
+
+		d_hash_rec_decref(&dfp->dfp_cont_table, &dfc->dfs_entry);
+		entry.attr.st_ino = ie->ie_stat.st_ino;
+		entry.entry_timeout = dfc->dfs_attr_timeout;
+		entry.generation = 1;
+		entry.ino = entry.attr.st_ino;
+		DFUSE_REPLY_ENTRY(ie, req, entry);
+		return;
+	}
 
-alloc_ie:
 	D_ALLOC_PTR(ie);
 	if (!ie)
 		D_GOTO(close, rc = ENOMEM);
 
-	ie->ie_root = true;
-
 	DFUSE_TRA_UP(ie, parent, "inode");
 
-	rc = dfs_lookup(dfs->dfs_ns, "/", O_RDWR, &ie->ie_obj, NULL,
+	rc = dfs_lookup(dfc->dfs_ns, "/", O_RDWR, &ie->ie_obj, NULL,
```

Comment:
Above, there is `d_hash_rec_find(&fs_handle->dpi_iet, ...)`. And if found, we copy in the inode entry. So this block is what happens if `!rlink`. But we're looking up root "/"? And we aren't inserting anything? I don't fully see the connection between the hash find above and this block of code below.

Reply:
Yes, this is for lookup where the result is the container.  The d_hash_rec_find() is looking up the inode for the container, if it's found (it's in cache) then it's update with the ostat and the function returns.  This particularly code-path is called the first time the container is opened only, so it does a lookup of / to find the root entry, later on this calls dfuse_reply_entry() which does a find_insert() of the inode where the insert will happen.

Reply:
Ah, that makes sense!

Reply:
It depends on what it's looking up, but for containers/pools it does an initial find, then if that fails it allocates everything and does a find_insert(), for dfs objects they're cheaper and the inode hash table is bigger so it always allocates everything, does a single find_insert() and then either frees or re-uses as needed.

The _find for containers/pools is the common case and they're infrequently updated so it makes sense from a performance perspective, they're completely optional however and were the last bit of code I wrote for this PR so I could also check for leaks/cleanup in the find_insert() paths.  If _find doesn't get anything but find_insert() comes back with something then there are two threads doing lookups concurrently.